File: DeepspaceRuby/lib/Hangar.rb
#encoding: UTF-8

require_relative 'HangarToUI'

module Deepspace
    class Hangar

        @maxElements
        @weapons
        @shieldBoosters

        def initialize(capacity)
            @maxElements = capacity
            @weapons = Array.new
            @shieldBoosters = Array.new
        end

        def self.newCopy(h)
            copy = new(h.maxElements)
            h.shieldBoosters.each{ |sB|
                copy.addShieldBooster(sB)
            }
            h.weapons.each{ |w|
                copy.addWeapon(w);
            }
            return copy;
        end

        def getUIversion
            return HangarToUI.new(self)
        end

        private
        
        def spaceAvailable
            return (@shieldBoosters.length + @weapons.length) < @maxElements
        end

        public

        def addWeapon(w)
            if spaceAvailable
                @weapons.push(w)
                return true
            else
                return false
            end
        end

        def addShieldBooster(s)
            if spaceAvailable
                @shieldBoosters.push(s)
                return true
            else
                return false
            end
        end

        def maxElements
            @maxElements
        end

        def weapons
            @weapons
        end

        def shieldBoosters
            @shieldBoosters
        end

        def removeShieldBooster(s)
            if s < 0 || s >= @shieldBoosters.length
                return nil
            end
            return @shieldBoosters.delete_at(s)
        end

        def removeWeapon(w)
            if w < 0 || w >= @weapons.length
                return nil
            end
            return @weapons.delete_at(w)
        end

        def to_s
            string = "MaxElements: #{@maxElements}"
            if !@weapons.empty?
                string += "\nWeapons: "
                @weapons.each { |w|
                    string += " #{w.to_s},"
                }
            else
                string += "\nWeapons: vacío"
            end
            if !@shieldBoosters.empty?
                string += "\nShieldBoosters: "
                @shieldBoosters.each { |sB|
                    string += " #{sB.to_s},"
                }
            else
                string += "\nShieldBoosters: vacío"
            end            
            return string
        end
    end
end

File: deepspace/Hangar.java
/*
 * @author: Jesús García León
 * @file: Hangar
 */
package deepspace;

import java.util.ArrayList;

/**
 * @brief Hangar que contiene los potenciadores dde escudo y las armas
 */
public class Hangar {
    private ArrayList<Weapon> weapons = new ArrayList<>();
    private ArrayList<ShieldBooster> shieldBoosters = new ArrayList<>();
    private int maxElements;
    
    Hangar(int capacity){
        maxElements = capacity;
    }
    
    //Referencia?
    Hangar(Hangar h){
        maxElements = h.getMaxElements();
        weapons = h.getWeapons();
        shieldBoosters = h.getShieldBoosters();
    }
    
    HangarToUI getUIversion(){
        return new HangarToUI(this);
    }
    
    private boolean spaceAvailable(){
        return (shieldBoosters.size() + weapons.size()) < maxElements;
    }
    
    public boolean addWeapon(Weapon w){
        if (spaceAvailable()){
            weapons.add(w);
            return true;
        }
        else return false;
    }
    
    public boolean addShieldBooster(ShieldBooster s){
        if(spaceAvailable()){
            shieldBoosters.add(s);
            return true;
        }
        else return false;
    }
    
    public int getMaxElements(){
        return maxElements;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters(){
        return shieldBoosters;
    }
    
    public ArrayList<Weapon> getWeapons(){
        return weapons;
    }
    
    public ShieldBooster removeShieldBooster(int s){
        if(s < 0 || s >= shieldBoosters.size())
            return null;
        else{
            ShieldBooster aux = shieldBoosters.get(s);
            shieldBoosters.remove(s);
            return aux;
        }
    }
    
    public Weapon removeWeapon(int w){
        if(w<0 || w >= weapons.size())
            return null;
        else{
            Weapon aux = weapons.get(w);
            weapons.remove(w);
            return aux;
        }
    }
    
    @Override
    public String toString() {
        String aux = "Espacio total: " + maxElements;
        aux += "\nArmas:\n";
        for (Weapon w: weapons)
            aux += w.toString() + "\n";
        aux += "Escudos:\n";
        for (ShieldBooster s: shieldBoosters)
            aux += s.toString() + "\n";
        return aux;
    }

}
